op-batcher/batcher/driver_test.go

```diff
@@ -116,3 +119,179 @@ func TestBatchSubmitter_SafeL1Origin_FailsToResolveRollupClient(t *testing.T) {
+		// expectations
+		expected     SyncActions
+		expectedLogs []string
+	}
+
+	testCases := []TestCase{
+		{name: "empty sync status",
+			newSyncStatus: eth.SyncStatus{},
+			expected:      SyncActions{waitForNodeSync: true},
+			expectedLogs:  []string{"empty sync status"},
+		},
+		{name: "sequencer restart",
+			newSyncStatus: eth.SyncStatus{
+				HeadL1:    eth.BlockRef{Number: 2},
+				CurrentL1: eth.BlockRef{Number: 1},
+			},
+			prevCurrentL1: eth.BlockRef{Number: 2},
+			expected:      SyncActions{waitForNodeSync: true},
+			expectedLogs:  []string{"sequencer currentL1 reversed"},
+		},
+		{name: "L1",
+			newSyncStatus: eth.SyncStatus{
+				HeadL1:    eth.BlockRef{Number: 2},
+				CurrentL1: eth.BlockRef{Number: 1},
+				SafeL2:    eth.L2BlockRef{Number: 100},
+				UnsafeL2:  eth.L2BlockRef{Number: 109},
+			},
+			prevCurrentL1: eth.BlockRef{Number: 1},
+			blocks:        queue.Queue[*types.Block]{block102, block103}, // note absence of block101
+			channels:      []ChannelStatuser{channel103},
+			expected: SyncActions{
+				clearState:   &eth.BlockID{},
+				blocksToLoad: [2]uint64{101, 109},
+			},
+			expectedLogs: []string{"new safe head is behind oldest block in state"},
+		},
+		{name: "batcher restart",
+			newSyncStatus: eth.SyncStatus{
+				HeadL1:    eth.BlockRef{Number: 2},
+				CurrentL1: eth.BlockRef{Number: 2},
+				SafeL2:    eth.L2BlockRef{Number: 104},
+				UnsafeL2:  eth.L2BlockRef{Number: 109},
+			},
+			prevCurrentL1: eth.BlockRef{Number: 1},
+			blocks:        queue.Queue[*types.Block]{block101, block102, block103},
+			channels:      []ChannelStatuser{channel103},
+			expected: SyncActions{
+				clearState:   &eth.BlockID{},
+				blocksToLoad: [2]uint64{105, 109},
+			},
+			expectedLogs: []string{"safe head above unsafe head"},
+		},
+		{name: "safe chain reorg",
+			newSyncStatus: eth.SyncStatus{
+				HeadL1:    eth.BlockRef{Number: 2},
+				CurrentL1: eth.BlockRef{Number: 2},
+				SafeL2:    eth.L2BlockRef{Number: 103, Hash: block101.Hash()}, // note hash mismatch
+				UnsafeL2:  eth.L2BlockRef{Number: 109},
+			},
+			prevCurrentL1: eth.BlockRef{Number: 1},
+			blocks:        queue.Queue[*types.Block]{block101, block102, block103},
+			channels:      []ChannelStatuser{channel103},
+			expected: SyncActions{
+				clearState:   &eth.BlockID{},
+				blocksToLoad: [2]uint64{104, 109},
+			},
+			expectedLogs: []string{"safe chain reorg"},
+		},
+		{name: "failed to make expected progress",
+			newSyncStatus: eth.SyncStatus{
+				HeadL1:    eth.BlockRef{Number: 2},
+				CurrentL1: eth.BlockRef{Number: 2},
+				SafeL2:    eth.L2BlockRef{Number: 101, Hash: block101.Hash()},
+				UnsafeL2:  eth.L2BlockRef{Number: 109},
+			},
+			prevCurrentL1: eth.BlockRef{Number: 1},
+			blocks:        queue.Queue[*types.Block]{block101, block102, block103},
+			channels:      []ChannelStatuser{channel103},
+			expected: SyncActions{
+				waitForNodeSync: true,
+				clearState:      &eth.BlockID{},
```

Comment:
None of these test cases returns a non empty value for `clearState`. Does that mean we need more coverage? Are the test assertions definitely working correctly?

Reply:
Why do so many of theses cases return an empty / zero valued but non nil `*eth.BlockID{}`? When would we want to do that? (It would trigger the batcher to clear its state entirely, and then submit a transaction as soon as possible afterward because it would mark the channel full immediately).

Reply:
It almost never makes sense to use an empty value for this.

Reply:
The problem is that the test is passing incomplete mocked data. We should fix that, and the assertions.

Reply:
[b8a43b5](https://github.com/ethereum-optimism/optimism/pull/13060/commits/b8a43b50b85410d6c216169cfbdaa81dff4511af)